README.md
# nys_parole_scraper

A package for scraping parole information from the NYS DOCCS public site.

Given an input of a csv files with identifying information of individuals (NYSID (New York state ID) or Name and DOB), this package will search for them via the [New York State DOCCS parolee lookup](https://publicapps.doccs.ny.gov/ParoleeLookup/default) and return all parole information, and summary statistics, for the individuals who are found in the system. 

## Use Case
This package was created to aid in the data collection efforts of service providers working with justice-involved populations in New York State. Organizations that are in need of parole information of their clients/participants to either further their own program design and evaluation or to successfully report to funders can use this package to do so. 

## Before you start

### Make sure you have Firefox installed as a browser on your computer
This package depends on using a headless Firefox browser. Make sure Firefox is installed on your computer before proceeding. Click [here](https://www.mozilla.org/en-US/firefox/new/?utm_medium=paidsearch&utm_source=google-rsa&utm_campaign=firefox-for-families&utm_content=A144_A203_301421&gclid=Cj0KCQiAqOucBhDrARIsAPCQL1Y1mhZc-bWiI1LXtW6IwfMGvn0WNIB4EV4x_A6z_9Sc8DHNEjDXL3YaAud1EALw_wcB) to download if you don't have it already. 

### Install Geckodriver and add to your Path variables
This package depends on installing Geckodriver and adding it your PATH variables in order to use Selnium for webscraping. Download Geckodriver [here](https://github.com/mozilla/geckodriver/releases) and see [here](https://www.softwaretestinghelp.com/geckodriver-selenium-tutorial/) for more information about Geckodriver and instructions on adding it to your PATH variables. 


## Installation

```bash
$ pip install nys_parole_scraper
```

## Usage

### Input 
This package requires that users provide identifying information that tells the scraper who to search the DOCCS parole database for. 

The input must be a .CSV file, and can live anywhere in you files. The .CSV file should must include the following columns, though they can be in any order:
- Unique ID (optional)
- NYSID (optional if first name, last name, and DOB provided)
- First Name (optional if accurate NYSID is provided)
- Last Name (optional if accurate NYSID is provided)
- DOB (optional if accurate NYSID is provided)

"Unique ID" refers to any unique identification system that your organization may use to identify clients/participants. This is meant to more easily link the returned data to yor own database. It is optional, and if some or all participants do not have a unique ID provided it will not impact the scraping. 

#### See below for sample .CSV setups: 

The scraper would search the DOCCS Parolee Lookup for all of the individuals shown below. Where the NYSID is missing, the first name, last name, and DOB are all provided. Where the DOB is missing, the NYSID is provided. And finally, all NYSIDs have the potential to be accurate, barring any data entry errors (NYSIDs are composed of 8 digits followed by one letter. If NYSIDs provided in the .CSV are not in this format, they will not be searched for as they are not accurate NYSID numbers.) Note that the UNique ID is optional, and the last person listed does not have one - that's okay!

| Unique ID      | NYSID | First Name     | Last Name | DOB |
| ----------- | ----------- |----------- | ----------- | ----------- |
| 19947      | 12345678A       | John     |  Doe        |    |
| 566657   | 12345678B        |  Jane       |    Doe     | 02/04/2008  |
| 928737      |               |   John      |   Smith     | 02/04/2008   |
|          | 12345678C        |   Jane       |   Smith    | 02/04/2008  |


Of the individuals in the sample below, only the last individual, Jane Smith, would be searched for. The first individual has an invalid NYSID composed of only 6 numbers followed by a letter. The second individual listed has a NYSID of "00000009", which is an invalid NYSID sometimes temporarily given to unknown individuals in custody. The third individual listed has no NYSID, but does not have completed first name, last name, and DOB lsited.  

| Unique ID      | NYSID | First Name     | Last Name | DOB |
| ----------- | ----------- |----------- | ----------- | ----------- |
| 19947      | 123456A       | John     |  Doe        |    |
| 566657   | 000000009        |  Jane       |    Doe     | 02/04/2008  |
| 928737      |               |   John      |        | 02/04/2008   |
| 884921   | 12345678C        |   Jane       |   Smith    | 02/04/2008  |


### Using the package

There are two potential ways to use the package. 

If you are only interested in the CSV and Excel output of the scraped parole data and the summary statistics, you can do as follows: 

```
from nys_parole_scraper import nys_parole_scraper

file_path = "C:/Users/parole_scraping/client_info.csv"
dir = "C:/Users/parole_scraping/Output_Folder"

nys_parole_scraper.parole_scraper(file_path, dir)
```
This will export a CSV file with all scraped data collected and an Excel file with summary statistics to a new timestamped folder in the designated directory. Output will also be returned in the console. 

If you want to explore, analyze, or modify the data  using python, and would like the output returned as python objects, you can assign the function to two variables, the first being the scraped output and the second being a list that will contain the summary statistic dataframes. For example: 

```
from nys_parole_scraper import nys_parole_scraper

file_path = "C:/Users/parole_scraping/client_info.csv"
dir = "C:/Users/parole_scraping/Output_Folder"

full_output, stats_list = nys_parole_scraper.parole_scraper(file_path, dir)
```

#### freq_table
While it is not the primary use of the nys_parole_scraper package, there is another function available for use on the full_output dataframe: freq_table(), available in the scraper_functions module of the nys_parole_scraper package, which allows for quick generation of any categorical variables in the output data. 

Parameters: freq_table(df, column, col_name)
- df: the pandas DataFrame containing the scraped parole output. In the example above, this is "full_output"
- column: a string; the name of the categorical column you would like to create a frequency table for
- index_name: a string; what you would like to name the column of categories

```
from nys_parole_scraper import scraper_functions as sf

conv_1_class_freq = sf.freq_table(full_output, "Class 1", "Conviction 1 Classes")
```


### parole_scraper Output Descriptions

##### The Parole Data
The scraped parole data, which will upload to your generated Output directory in the directory path provided as a parameter, will include the following information scraped from the DOCCS Parole Lookup site. 
- ID: the unique ID provided by the package user in the csv file input; blank if not provided
- DIN (Department Identification Number): a unique ID used by DOCCS
- Name
- Date of Brith
- Age (calculated by the package; not directly available on DOCCS site)
- Race/ethnicity
- Release to Parole supervision (date)
- Months Since Release (calculated by the package; not directly available on DOCCS site)
- Parole Status
- Effective Date
- Senior Parole Officer
- Parole Officer
- Office (of P.O)
- Street (of P.O)
- City (of P.O)
- Phone (of P.O)
- Crime of Conviction 1 through Crime of Conviction 10 
- Class 1 through 10 (class of conciviciton charge)
- County 1 through 10 (county of conviction)
- Date Info Scraped (the date of package use; not available on DOCCS site)

Where a value is blank for an individual, that information was not available on the individual's DOCCS parole page. 

for more detailed information and definitions of the data provided by the DOCCS Parole Lookup, please see the [DOCCS Parolee Lookup Glossary.](https://publicapps.doccs.ny.gov/ParoleeLookup/About?form=glossary)


##### The Summary Statistics
The summary statistics Excel file will contain sheets with the following names/data: 

Summary Statistics: 
- summary statistics (min, mean, and max) of continuous variables age, months since release to parole (from current date of package run), and total convictions

Age: 
- Frequency table (count and %) of age categories: < 20, 21-30, 31-40, 41-50, 51-60 > 60 

Race.Ethnicity: 
- Frequency table (count and %) of race/ethnicity

Parole Status: 
- Frequency table (count and %) of current parole status. Categories may include, but are not limited to: discharged, active, in custody,  revoked, absconded, and deceased.

Unique people per County
- Frequency table (count and %) of unique individuals with a conviction in the county. Note that because individuals may have multiple convictions, all in different counties, and that this is counting the unique individuals with a conviction in each county rather than the count of counties out of total convictions, this table will likely add to more than 100%.

Top charge
- Frequency table (count and %) of top charge (crime of convction 1)

Unique People per Charge Type
- Frequency table (count and %) of unique individuals per conviction charge type (e.g.: attempted robbery in the 2nd). Note that because individuals may have multiple convictions, and that this is counting the unique individuals with a specific charge type rather than the count of each charge type out of all convictions, this table will likely add to more than 100%.

## Read the Docs

https://nys-parole-scraper.readthedocs.io/en/latest/

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`nys_parole_scraper` was created by Kellyann Hayes. It is licensed under the terms of the MIT license.

## Credits

`nys_parole_scraper` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (15/12/2022)

- First release of `nys_parole_scraper`!

tests/test_nys_parole_scraper.py
import sys
sys.path.append("C:/Users/khaye/mds_course/src/mds/Hayes_Kellyann/Final_Project/nys_parole_scraper")

from nys_parole_scraper import scraper_functions
from nys_parole_scraper import nys_parole_scraper
import pandas as pd
from pandas.testing import assert_frame_equal
import pytest

#==============================================================================
# TEST FREQ_TABLE
#=============================================================================

# fake data
data = {'Animal':['Cat', 'Dog', 'Cat', 'Dog', 'Bird'],
        'Color':['Grey', 'Brown', 'Black', 'Brown', 'White'],
        'Weight' : [1,2,3.4,4,1]
        }
df = pd.DataFrame(data)

# freq_table test function
def test_freq_table():
    expected = pd.DataFrame({'Animal':['Cat','Dog','Bird'],'Count':[2, 2, 1],'%':['40.0%', '40.0%', '20.0%']}) 
    actual = scraper_functions.freq_table(df, 'Animal', 'Animal')
    assert_frame_equal(expected, actual)
    
# freq_table test function of paramter 1
def test_param1_int():
    with pytest.raises(ValueError):
        scraper_functions.freq_table(1, 'animal', 'Animal')
        
# freq_table test function of paramter 1
def test_param1_str():
    with pytest.raises(ValueError):
        scraper_functions.freq_table("Animal", 'animal', 'Animal')
    
# freq_table test function of paramter 1
def test_param2_int():
    with pytest.raises(ValueError):
        scraper_functions.freq_table(df, 1, 'Animal')
        
# freq_table test function of paramter 2
def test_param3_int():
    with pytest.raises(ValueError):
        scraper_functions.freq_table(df, "animal", 1)
    
    
    
    
#==============================================================================
# TEST PAROLE_SCRAPER
#=============================================================================
    
# test function for ValueError on parole scraper
def test_valerror1():
    with pytest.raises(ValueError):
        nys_parole_scraper.parole_scraper(1, "C:/test_path")
        
# test function for ValueError on parole scraper
def test_valerror2():
    with pytest.raises(ValueError):
        nys_parole_scraper.parole_scraper(1.56, "C:/test_path")
        
# test function for ValueError on parole scraper
def test_valerror3():
    with pytest.raises(ValueError):
        nys_parole_scraper.parole_scraper("C:/test_path", 1)

# test function for ValueError on parole scraper
def test_valerror4():
    with pytest.raises(ValueError):
        nys_parole_scraper.parole_scraper("C:/test_path", 1.56)

src/nys_parole_scraper/scraper_functions.py
# -*- coding: utf-8 -*-
"""
@author: khayes
"""
import pandas as pd
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC

#=============================================================================
# FUNCTION: SCRAPE TABLE 1
#=============================================================================

def scrape_table1(driver, df):
    #global df
    ##Table 1 - "ParoleeInformation"
    tbl1= driver.find_element(By.XPATH,
                              "//*[@id='MainContent_paroleeInformation']"
                             ).get_attribute('outerHTML')
    df  = pd.read_html(tbl1)
    df = df[0]
    df = df.transpose()
    df.columns = df.iloc[0]
    df = df[1:]
    return df

#=============================================================================
# FUNCTION: SCRAPE TABLE 2
#=============================================================================

def scrape_table2(driver, df):
    ##If Table 2 (supervision info) exists, scrape:
    tbl2= driver.find_element(By.XPATH,
        "//*[@id='MainContent_supervisionInformation']"
        ).get_attribute('outerHTML')
    df2  = pd.read_html(tbl2)
    df2 = df2[0]
    df2 = df2.transpose()

    df2.columns = df2.iloc[0]
    df2 = df2[1:]
    df = pd.merge(df, df2, how="cross")
    return df

#=============================================================================
# FUNCTION: SCRAPE TABLE 3
#=============================================================================

def scrape_table3(driver, df, uid):
    #if offense table exists:
    tbl3= driver.find_element(By.XPATH,
        "//*[@id='MainContent_offenseInformationTable']"
        ).get_attribute('outerHTML')
    df3  = pd.read_html(tbl3)
    df3 = df3[0]
    df3 = df3.transpose()
    df3 = df3.reset_index()
    df3 = df3.drop(columns=['index'])
    
    dic = {0: [], 1: [], 
           2: [], 3: [], 
           4: [], 5: [],
           6: [], 7: [],
           8: [], 9: []}
    
    ################## Row 1 - Charges
    df3_row1 = pd.DataFrame(data = dic)
    
    df3_row1 = df3_row1.append(df3.iloc[[0]])
    
    df3_row1.rename(columns={0: 'Crime of conviction 1', 
                             1: 'Crime of conviction 2',
                             2: 'Crime of conviction 3',
                             3: 'Crime of conviction 4',
                             4: 'Crime of conviction 5',
                             5: 'Crime of conviction 6',
                             6: 'Crime of conviction 7',
                             7: 'Crime of conviction 8',
                             8: 'Crime of conviction 9',
                             9: 'Crime of conviction 10'}, inplace=True)
    
    df = pd.merge(df, df3_row1, how="cross")
    
    #################### Row 2 - Classes
    df3_row2 = pd.DataFrame(data = dic)
    
    df3_row2 = df3_row2.append(df3.iloc[[1]])
    
    df3_row2.rename(columns={0: 'Class 1', 
                             1: 'Class 2',
                             2: 'Class 3',
                             3: 'Class 4',
                             4: 'Class 5',
                             5: 'Class 6',
                             6: 'Class 7',
                             7: 'Class 8',
                             8: 'Class 9',
                             9: 'Class 10'}, inplace=True)
    
    df = pd.merge(df, df3_row2, how="cross")
    
    ######################## Row 3 - Boroughs
    df3_row3 = pd.DataFrame(data = dic)
    
    df3_row3 = df3_row3.append(df3.iloc[[2]])
    
    df3_row3.rename(columns={0: 'County 1', 
                             1: 'County 2',
                             2: 'County 3',
                             3: 'County 4',
                             4: 'County 5',
                             5: 'County 6',
                             6: 'County 7',
                             7: 'County 8',
                             8: 'County 9',
                             9: 'County 10'}, inplace=True)
    
    df = pd.merge(df, df3_row3, how="cross")
    #df["parid"] = uid
    df.insert(loc = 0,
          column = 'ID',
          value = uid)
    return df


#==========================================================
# DEFINE NEW SEARCH
#==========================================================
def new_search(driver, wait):
    wait.until(EC.visibility_of_element_located((
        By.XPATH, "//*[@id='MainContent_NewSearch']"))) 
    driver.find_element(By.XPATH, "//*[@id='MainContent_NewSearch']").click()
    
    
#==========================================================
# DEFINE FREQUENCY TABLE FUNCTION
#==========================================================
def freq_table(df, column, col_name):
    """
    Create a  frequency table.
    
    Parameters
    ----------
    df : pandas DataFrame
        the dataFrame witht he data you would like to create a frequency table of
        
    column : String
        The column in the dataframe you would like to make a frequency table of.
        
    col_name : String
        The name you would like to give to the first column of the returned DataFrame


    Returns
    -------
    To return the following python objects as DataFrames, assign the function to 
    two variables (ex: df1, df2 = parole_scraper(file_path, directory)). To only export 
    the dataframes to the provided directory, you do not need to assign the 
    function to variables (ex: parole_scraper(file_path, directory))
    
    df : console output or pandas DataFrame if assigned to variable
                
    Example
    -------- 
    Returning DataFrames object: 
    >>> from nys_parole_scraper import scraper_functions as sf
    >>> data = {'animal':['Cat', 'Dog', 'Cat', 'Dog', 'Bird'],
            'color':['Grey', 'Brown', 'Black', 'Brown', 'White'],
            'weight' : [1,2,3.4,4,1]
            }
    >>> df = pd.DataFrame(data)
    >>> animal_frequency = sf.freq_table(df, 'animal', 'Animals')
    animal_frequency
    
    Animals | Count  |   %
    ------------------------
    Cat     |   2    | 40.0%
    Dog     |   2    | 40.0%
    Bird    |   1    | 20.0%
    
    
    Returning frequency table only in console output: 
    >>> from nys_parole_scraper import scraper_functions as sf
    >>> data = {'animal':['Cat', 'Dog', 'Cat', 'Dog', 'Bird'],
            'color':['Grey', 'Brown', 'Black', 'Brown', 'White'],
            'weight' : [1,2,3.4,4,1]
            }
    >>> df = pd.DataFrame(data)
    >>> sf.freq_table(df, 'animal', 'Animals')
    
    Animals | Count  |   %
    ------------------------
    Cat     |   2    | 40.0%
    Dog     |   2    | 40.0%
    Bird    |   1    | 20.0%
    
    """
    if not isinstance(df, pd.DataFrame):
        raise ValueError('column argument must be a string')
    if not isinstance(column, str):
        raise ValueError('column argument must be a string')
    if not isinstance(col_name, str):
        raise ValueError('col_name argument must be a string')
    c = df[column].value_counts(dropna=False)
    p = df[column].value_counts(dropna=False, normalize=True).mul(100).round(1).astype(str) + '%'
    df_new = pd.concat([c,p], axis=1, keys=['Count', '%'])
    df_new.index.rename(col_name, inplace=True) 
    df_new.reset_index(inplace = True)
    return df_new


src/sample.py
# -*- coding: utf-8 -*-
"""
Created on Wed Dec 14 20:27:32 2022

@author: khaye
"""

from parole_scraper_MDS_4 import parole_scraper
from datetime import datetime
from datetime import date
import numpy as np 
import pandas as pd
from test_freq import freq_table
import os, sys
now = date.today()
today_str = now.strftime("%m/%d/%Y")
today = datetime.strptime(today_str, "%m/%d/%Y")


path="C:/Users/khaye/OneDrive/Documents/Parole_Scraping/DATA_FILES/test_data.csv"
dire = "C:/Users/khaye/OneDrive/Documents/Parole_Scraping"
path="test"
dire = "blah blah"
os.path.isfile(path)

output, stats = parole_scraper(path, dire)

if os.path.isfile(path) == False:
    raise ValueError('file_path argument must be a path to an existing file. Please check your input.')
    
os.path.isdir(dire)

os.chir
    
except False: 
    raise ValueError('file_path argument must be a path to an existing file. Please check your input.')


nysids_csv.style.set_caption("NYSIDS DF")
nysids_csv

full_output[["Date of birth:", "Release to parole supervision:"]]= full_output[["Date of birth:", "Release to parole supervision:"]].astype('datetime64[ns]')

#full_output["Age"] = (today - full_output["Date of birth:"]) / np.timedelta64(1, 'Y')
#full_output["Age2"] = np.floor(full_output["Age"])

full_output["Months Since Release to Parole"] = (today - full_output["Release to parole supervision:"]) / np.timedelta64(1, 'M')

full_output["Months Since Release to Parole"] = np.around(full_output["Months Since Release to Parole"], decimals=1, out=None)

full_output["Date Info Scraped:"] = today_str


idob = full_output.columns.get_loc("Date of birth:")
age = (today - full_output["Date of birth:"]) / np.timedelta64(1, 'Y')
age = np.floor(age)

full_output.insert(loc = idob +1,
          column = 'Age:',
          value = age)

irel = full_output.columns.get_loc("Release to parole supervision:")
mrelease = (today - full_output["Release to parole supervision:"]) / np.timedelta64(1, 'M')
mrelease = np.around(mrelease, decimals=1, out=None)

full_output.insert(loc = irel +1,
          column = "Months Since Release:",
          value = mrelease)

convictions_list = ['Crime of conviction 1',
           'Crime of conviction 2', 
           'Crime of conviction 3', 
           'Crime of conviction 4', 
           'Crime of conviction 5',
           'Crime of conviction 6', 
           'Crime of conviction 7', 
           'Crime of conviction 8', 
           'Crime of conviction 9', 
           'Crime of conviction 10']

iconv = full_output.columns.get_loc("Crime of conviction 10") 
count_conv = full_output[[convictions_list]].apply(lambda x: x.notnull().sum(), axis='columns')
full_output.insert(loc = iconv +1,
          column = "Total Convictions",
          value = count_conv)

describe = full_output.describe()

from pandas_profiling import ProfileReport
profile = ProfileReport(full_output, title='Profiling Report', explorative=False)

stats_numeric = full_output[["Age:", "Months Since Release:","Total Convictions"]].describe().loc[['count','min','mean','max']]


# Frequency tables for each categorical feature
crosstab = pd.DataFrame()
for column in full_output[["Race / ethnicity:", "Parole status:", "Senior Parole Officer:"]].select_dtypes(include=['object']).columns:
    table = pd.crosstab(index=full_output[column], columns='% observations', normalize='columns')*100
    crosstab.append(table)

c = full_output["Race / ethnicity:"].value_counts(dropna=False)
p = full_output["Race / ethnicity:"].value_counts(dropna=False, normalize=True).mul(100).round(1).astype(str) + '%'
race = pd.concat([c,p], axis=1, keys=['Count', '%'])

# def freq_table(df, column):    
#     c = df[column].value_counts(dropna=False)
#     p = df[column].value_counts(dropna=False, normalize=True).mul(100).round(1).astype(str) + '%'
#     pd.concat([c,p], axis=1, keys=['Count', '%'])







output_stats = full_output.copy()



# create age bins
bins = [0,20,30,40,50,60,150]
group_names=['0 - 20','21-30','31-40','41-50','51-60','60 +']
output_stats['age_bin']=pd.cut(output_stats['Age:'],bins,labels=group_names).astype(object)

county_list = ['County 1',
           'County 2', 
           'County 3', 
           'County 4', 
           'County 5',
           'County 6', 
           'County 7', 
           'County 8', 
           'County 9', 
           'County 10']

output_stats[county_list] = output_stats[county_list].fillna("")

counties = np.unique(output_stats[county_list].values)
counties = [item for item in counties if item != ""]

county_unique_freq = pd.DataFrame(columns=['Count', '%'], index = counties)
for county in counties:
    output_stats[county+'_unique'] = output_stats[['County 1',
            'County 2', 
            'County 3', 
            'County 4', 
            'County 5',
            'County 6', 
            'County 7', 
            'County 8', 
            'County 9', 
            'County 10']].apply(
                lambda x: x.str.contains(county ,case=False)).any(axis=1).astype(int)
    c = output_stats[county+'_unique'].sum()
    p = str(round((c/len(output_stats.index))*100, 1))+ '%'
    county_unique_freq.loc[county].Count = c
    county_unique_freq.loc[county]['%'] = p



# Yes/No for parole officer info
output_stats['Officer Info Provided'] = np.where(output_stats['Senior Parole Officer:'].notnull, 'Yes', 'No')


# create a list of the values we want to assign for each condition
values = ['tier_4', 'tier_3', 'tier_2', 'tier_1']

# create a new column and use np.select to assign values to it using our lists as arguments
#df['tier'] = np.select(conditions, values)


    # new = pd.concat([c,p], axis=1, keys=['Count', '%'])
    # full_output = pd.concat(county_unique_freq, new)

# for col in output_stats[["Race / ethnicity:", "Parole status:", "Senior Parole Officer:"]]
# c = df[column].value_counts(dropna=False)
# p = df[column].value_counts(dropna=False, normalize=True).mul(100).round(1).astype(str) + '%'
# df_new = pd.concat([c,p], axis=1, keys=['Count', '%'])            


race_freq = freq_table(output_stats, "Race / ethnicity:")
pstatus_freq = freq_table(output_stats, "Parole status:")
age_freq = freq_table(output_stats, "age_bin")
officer_freq = freq_table(output_stats, 'Officer Info Provided')
top_charge_freq = freq_table(output_stats, 'Crime of conviction 1')


full_output = pd.read_csv("C:/Users/khaye/OneDrive/Documents/Parole_Scraping/Output_20221215_18.48.32/parole_full_output_20221215_18.48.32.csv", delimiter=',', dtype='str')
top_charge_freq = freq_table(full_output, 'Crime of conviction 1')
# Create a Pandas Excel writer
writer = pd.ExcelWriter(dire + '/summary_statistics.xlsx', engine='xlsxwriter')

# Write each dataframe to a different worksheet.
stats_numeric.to_excel(writer, sheet_name='Summary Statistics')
age_freq.to_excel(writer, sheet_name='Age')
race_freq.to_excel(writer, sheet_name='Race.Ethnicity')
pstatus_freq.to_excel(writer, sheet_name='Parole Status')
officer_freq.to_excel(writer, sheet_name='P.O. Info Provided')
top_charge_freq
# Close the Pandas Excel writer and output the Excel file.
writer.save()




# get counts of charges
conv_list = ['Crime of conviction 1',
           'Crime of conviction 2', 
           'Crime of conviction 3', 
           'Crime of conviction 4', 
           'Crime of conviction 5',
           'Crime of conviction 6', 
           'Crime of conviction 7', 
           'Crime of conviction 8', 
           'Crime of conviction 9', 
           'Crime of conviction 10']

output_stats[conv_list] = output_stats[conv_list].fillna("")

convictions = np.unique(output_stats[conv_list].values)
convictions = [item for item in convictions if item != ""]

global convictions_freq
convictions_freq = pd.DataFrame(columns=['Count', '%'], index = convictions)
for conv in convictions:
    output_stats[conv+'_unique'] = output_stats[['Crime of conviction 1',
            'Crime of conviction 2', 
            'Crime of conviction 3', 
            'Crime of conviction 4', 
            'Crime of conviction 5',
            'Crime of conviction 6', 
            'Crime of conviction 7', 
            'Crime of conviction 8', 
            'Crime of conviction 9', 
            'Crime of conviction 10']].apply(
                lambda x: x.str.contains(conv ,case=False)).any(axis=1).astype(int)
    c = output_stats[conv+'_unique'].sum()
    #p = str(round((c/len(output_stats.index))*100, 1))+ '%'
    convictions_freq.loc[conv].Count = c
    #convictions_freq.loc[conv]['%'] = p



c = output["Crime of conviction 1"].value_counts(dropna=False)
p = output["Crime of conviction 1"].value_counts(dropna=False, normalize=True).mul(100).round(1).astype(str) + '%'
df_new = pd.concat([c,p], axis=1, keys=['Count', '%'])
df_new.index.rename('Top Charge', inplace=True) 